enzy_htp/structure/structure_constraint/api.py

```diff
@@ -1,43 +1,552 @@
+    of the methods yield junk values since no measurement is needed for this Class.
+    """
+    
+    def __init__(self, atoms:List[Atom] ):
+        super().__init__(self, atoms, 0.0, dict())
+     
+    def is_cartesian_freeze(self) -> bool:
+        """Is this a cartesian freeze constraint? Always True for this class."""
+        return False
+
+    def correct_num_atoms(self) -> bool:
+        """True as long as there is at least one atom in the constraint."""
+        return len(self.atoms)
+
+    def score_energy(self) -> float:
+        """Always returns 0.0. Not needed for this type of constraint."""
+        return 0.0
+
+    def current_geometry(self) -> float:
+        """Always returns 0.0. Not needed for this type of constraint."""
+        return 0.0
+
+
+class DistanceConstraint(StructureConstraint):
+    """Specialization of StructureConstraint() for the distance between two Atom()'s."""
+    def is_distance_constraint(self) -> bool:
+        """Is this a distance constraint? Always True for this class."""
+        return True
+
+    def correct_num_atoms(self) -> bool:
+        """A distance constraint should have two atoms."""
+        return len(self.atoms) == 2
+
+    def current_geometry(self) -> float:
+        """The cartesian distance between two 3D points."""
+        return self.atoms[0].distance_to(self.atoms[1])
+
+
+class AngleConstraint(StructureConstraint):
+    """Specialization of StructureConstraint() for the angle between three Atom()'s."""
+    def is_angle_constraint(self) -> bool:
+        """Is this an angle constraint? Always True for this class."""
+        return True 
+
+    def correct_num_atoms(self) -> bool:
+        """An angle constraint should have three atoms."""
+        return len(self.atoms) == 3
+
+    def current_geometry(self) -> float:
+        """The angle between three 3D points using dot-product."""
+        a = np.array(self.atoms[0].coord)
+        b = np.array(self.atoms[1].coord)
+        c = np.array(self.atoms[2].coord)
+        
+        ba = a - b
+        bc = c - b
+        
+        cosine_angle = np.dot(ba, bc) / (np.linalg.norm(ba) * np.linalg.norm(bc))
+        return np.degrees(np.arccos(cosine_angle))
+        
+
+class DihedralConstraint(StructureConstraint):
+    """Specialization of StructureConstraint() for the dihedral angle between four Atom()'s"""
+    
+    def is_dihedral_constraint(self) -> bool:
+        """Is this a dihedral constraint?"""
+        return True 
+
+    def correct_num_atoms(self) -> bool:
+        """A dihedral constraint should have 4 atoms."""
+        return len(self.atoms) == 4
+
+    def current_geometry(self) -> float:
+        """Measurement for the current dihedral between the 4 atoms in the constraint."""
+        p0 = np.array(self.atoms[0])
+        p1 = np.array(self.atoms[1])
+        p2 = np.array(self.atoms[2])
+        p3 = np.array(self.atoms[3])
```

Comment:
We can add it to Atom too.

Also we should use `core.math_helper.get_dihedral()` for this

Reply:
This is for current_geometry()